### PR TITLE
SignalR: Reference shared framework

### LINKED ISSUE #27325: Reference shared framework in SignalR docs
I think we should cross-ref to https://learn.microsoft.com/aspnet/core/fundamentals/target-aspnetcore?view=aspnetcore-6.0&tabs=visual-studio from this page or multiple pages in SignalR. I see many questions where people are trying to update their SignalR package in class libraries (since we removed the package in 3.0+ and moved to the shared framework).

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 1387f556-b4b7-70f9-5bd0-f0bccc597689
* Version Independent ID: fb016cdc-d281-6b5b-7466-2c718e7274b4
* Content: [Use hubs in ASP.NET Core SignalR](https://learn.microsoft.com/en-us/aspnet/core/signalr/hubs?view=aspnetcore-6.0)
* Content Source: [aspnetcore/signalr/hubs.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/signalr/hubs.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-signalr**
* GitHub Login: @bradygaster
* Microsoft Alias: **bradyg**

---
[Associated WorkItem - 33242](https://dev.azure.com/msft-skilling/Content/_workitems/edit/33242)

### PR DESCRIPTION
**Pull request by wadepickett**
Fixes #27325

[Internal Review - SignalR assemblies in shared framework: 2.2-3.0 Migration](https://review.learn.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-8.0&branch=pr-en-us-28398&tabs=visual-studio#signalr-assemblies-in-shared-framework)
[Internal Review - Hubs.md .NET 7  ---Also moved latest version to top](https://review.learn.microsoft.com/en-us/aspnet/core/signalr/hubs?view=aspnetcore-7.0&branch=pr-en-us-28398#configure-signalr-hubs)
[Internal Review - Hubs.md - .NET 6](https://review.learn.microsoft.com/en-us/aspnet/core/signalr/hubs?view=aspnetcore-6.0&branch=pr-en-us-28398#configure-signalr-hubs-1)
[Internal Review - Hubs.md - .NET 3.1-5](https://review.learn.microsoft.com/en-us/aspnet/core/signalr/hubs?view=aspnetcore-5.0&branch=pr-en-us-28398#configure-signalr-hubs-2)
[Internal Review - hubcontext.md ](https://review.learn.microsoft.com/en-us/aspnet/core/signalr/hubcontext?view=aspnetcore-8.0&branch=pr-en-us-28398#additional-resources)
[Internal Review - api-design.md ](https://review.learn.microsoft.com/en-us/aspnet/core/signalr/api-design?view=aspnetcore-8.0&branch=pr-en-us-28398#additional-resources)

Adding Notes that SignalR is included in the Microsoft.AspNetCore.App shared framework rather than as NuGet packages.

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
This should go in the migration doc. This doc note can be far shorter with a link to the migration doc.

File: aspnetcore/signalr/hubs.md [hunk 1]
```diff
@@ -23,6 +23,9 @@ To register the services required by SignalR hubs, call <xref:Microsoft.Extensio
 
 :::code language="csharp" source="~/../AspNetCore.Docs.Samples/signalr/hubs/samples/6.x/SignalRHubsSample/Program.cs" id="snippet_AddSignalR" highlight="4":::
 
+> [!NOTE]
```
Comment (Rick-Anderson) [inline aspnetcore/signalr/hubs.md:26]: Why is this only in .NET 6? Why is .NET 7 at the bottom, it needs to be on top.

**Comment by Rick-Anderson** [comment]
@wadepickett  review URL?

File: aspnetcore/signalr/hubs.md [hunk 2]
```diff
@@ -23,6 +23,9 @@ To register the services required by SignalR hubs, call <xref:Microsoft.Extensio
...
+> With the release of .NET Core 3.0, many ASP.NET Core assemblies, including those of SignalR, are no longer published to NuGet as packages. Instead, the assemblies are included in the Microsoft.AspNetCore.App shared framework, which is installed with the .NET Core SDK and runtime installers. See [Use the ASP.NET Core shared framework](xref:fundamentals/target-aspnetcore#use-the-aspnet-core-shared-framework) for more information.
+
```
Comment (Rick-Anderson) [inline aspnetcore/signalr/hubs.md:28]: Do we really want to have this first thing at the top of the article? The Note is too much bling.

```suggestion
SignalR no longer depends on getting many assemblies from NuGet, those assemblies are contained in the `Microsoft.AspNetCore.App` shared framework. For more information, see [Microsoft.AspNetCore.App replaces many NuGet packages](xref:the/right/migration/doc#H2).
```

I recommend putting this at the bottom of the file. Experienced SignalR dev's migrating aren't going to look in this doc, they'll look in the migration doc. OK to put a link at the bottom tho

**Comment by wadepickett** [comment]
@Rick-Anderson, this is a draft.  I promise I will ask for a review when it is ready for a review. 
image

**Comment by Rick-Anderson** [comment]
Right, but it's not like you would have made those changes when you were ready for a review.

**Comment by wadepickett** [comment]
OK, I understand.  You don't think I would have made those changes before a review, and you were trying to save me some time.

File: aspnetcore/signalr/hubs.md (hunk 1 above)
Comment (wadepickett) [reply aspnetcore/signalr/hubs.md:26]: **"Why is this only in .NET 6? Why is .NET 7 at the bottom, .NET 7 needs to be on top."**
* Looking at the history, Brennan added 7.0 at the bottom in April.  
* Later you left .NET 7 in the bottom in [your update in the .NET 7 version in July] (https://github.com/dotnet/AspNetCore.Docs/commit/cc7a860efe9d3b866871ce83c5e533e67b36feec)

In my draft, I unfortunately I didn't notice the 7.0 at the bottom and thought 6.0 at the top was the latest and edited that.  I should have looked around further, however, I know better to watch out for them.  I have seen several topics where the latest was added to the bottom.  When I run into those, I would move them to the top.  As I will this one now.

File: aspnetcore/signalr/hubs.md (hunk 2 above)
Comment (wadepickett) [reply aspnetcore/signalr/hubs.md:28]: I agree.

**Comment by wadepickett** [comment]
OK, I understand.  You don't think I would have made those changes before a review, and you were trying to save me some time.

> @wadepickett review URL?

It is a draft.  It was not made available for review.  I drop that in as soon as I check the build and before putting up for review.

**Comment by Rick-Anderson** [comment]
> > @wadepickett review URL?
> 
> It is a draft. It was not made available for review. I drop that in as soon as I check the build and before putting up for review.

I prefer that you put it in after you get a build along with Fixes #xyz.

**Comment by wadepickett** [comment]
> > > @wadepickett review URL?
> > 
> > 
> > It is a draft. It was not made available for review. I drop that in as soon as I check the build and before putting up for review.
> 
> I prefer that you put it in after you get a build along with Fixes #xyz.
I understand your preference.
[#28339](https://github.com/dotnet/AspNetCore.Docs/pull/28339)

File: aspnetcore/signalr/hubs.md (hunk 2 above)
Comment (wadepickett) [reply aspnetcore/signalr/hubs.md]: @Rick-Anderson, I'll put it at the end of the topics as you ask but drop it back into a note.  That way it doesn't not look like it is part of the context of whatever the last section is, and a dedicated new section does not make sense of course.

**Comment by wadepickett** [comment]
re build

File: aspnetcore/signalr/hubs.md [hunk 3]
```diff
@@ -114,12 +116,140 @@ Using `Hub<IChatClient>` enables compile-time checking of the client methods. Th
...
+```csharp
+public class ChatHub : Hub
+{
+    public async Task<string> WaitForMessage(string connectionId)
+    {
+        var message = await Clients.Client(connectionId).InvokeAsync<string>(
+            "GetMessage");
+        return message;
+    }
+}
```
Comment (Rick-Anderson) [inline aspnetcore/signalr/hubs.md:134]: All this needs to be moved to a project that builds.

Comment (wadepickett) [reply aspnetcore/signalr/hubs.md:134]: Thanks, I will take care of that in #28459.   Out of scope here for this link inclusion.

**Comment by wadepickett** [comment]
@Rick-Anderson, I did not put the note in all advanced topics, which seemed too broad.  I put the note in the following topics that seemed like a good place and context to remind the target audience who might be updating their existing libraries.  :

- signalr/api-design.md   
- signalr/configuration.md
- signalr/hub-filters.md  
- signalr/hubcontext.md   
- signalr/hubs.md
- signalr/streaming.md 

I specified _**_ASP.NET Core_**_ SignalR where otherwise I would be adding a new version moniker for a few lines of text.  I felt like adding more little moniker chunks in some of those old topics where there were no moniker versioning to begin with was going to make it needlessly messy for future updates.

There were a couple of topics where I needed to pull the latest version to the top.   So you will have to search for the phrase "SignalR no longer depends" to see what I added.  Normally we would move versions to the top first, merge and then edit so that would be cleaner, but they were being added afterwards.

**Comment by wadepickett** [comment]
...adding additional review links...

File: aspnetcore/signalr/api-design.md [hunk 4]
```diff
@@ -75,3 +75,6 @@ The older clients won't be expecting the `Sender` value, so they'll ignore it. A
 [!code-typescript[OnWithObjectNew](api-design/sample/Samples.ts?name=OnWithObjectNew&highlight=2-5)]
 
 In this case, the new client is also tolerant of an old server that doesn't provide the `Sender` value. Since the old server won't provide the `Sender` value, the client checks to see if it exists before accessing it.
+
+> [!NOTE]
+> ASP.NET Core SignalR no longer depends on getting many assemblies from NuGet, those assemblies are implicitly referenced in the `Microsoft.NET.Sdk.Web`. For more information, see [Differences between ASP.NET SignalR and ASP.NET Core SignalR](xref:signalr/version-differences#how-to-identify-the-signalr-version).
```
Comment (Rick-Anderson) [inline aspnetcore/signalr/api-design.md:80]: This needs to be an include

**Review by Rick-Anderson — CHANGES_REQUESTED** [review]

**Comment by wadepickett** [comment]
> @wadepickett see above where I wrote:
> 
> > This should go in the migration doc. This doc note can be far shorter with a link to the migration doc.

I made it shorter and linked to that topic which starts off by showing the differences between versions for packages.  I thought that is what you meant.   I guess I can shorten it from this single sentence:

"ASP.NET Core SignalR no longer depends on getting many assemblies from NuGet, those assemblies are implicitly referenced in the `Microsoft.NET.Sdk.Web`. "

to a shorter one like this, but to save those few words the reader for sure has to go to the link to see what is going on.  That doesn't feel better to me, but I can go that route.

ASP.NET Core SignalR no longer depends on getting many assemblies from NuGet, see...

File: aspnetcore/signalr/api-design.md (hunk 4 above)
Comment (wadepickett) [reply aspnetcore/signalr/api-design.md:80]: OK, I will move it into an include, thanks.

**Comment by Rick-Anderson** [comment]
Dan Roth likes to avoid history lessons. I'd be inclined to write it

ASP.NET Core SignalR doesn't depend on getting many assemblies from NuGet. For more information, see <link here>"

Can you provide a link to the relevant migration doc? 
As I previously stated, This only needs to be in a couple places and probably doesn't belong in onboarding docs. Folks  migrating:

1. Don't look in on boarding docs.
2. Look in the migration docs.

**Comment by wadepickett** [comment]
I did provide a link to the migration doc.  Maybe I don't understand what you are referring to when you say "the migration doc".

As far as I saw, the only thing we had that we referred to as "the migration" doc for signal R was the signalr/version-differences.md doc.  In  #6463 you ask for a new migration doc, Brennan says to use the signalr/version-differences.md, he updates that then that is "the migrations" doc for SignalR.     That is what I have been linking to since you asked to link to the migrations doc instead of what Brennan suggested to use.  I linked to here:  For more information, see [Differences between ASP.NET SignalR and ASP.NET Core SignalR](xref:signalr/version-differences#how-to-identify-the-signalr-version).

Is there some other doc you consider "the migrations doc" that you want me to link to?

**Comment by Rick-Anderson** [comment]
> [2.2 to 3.0](https://learn.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-7.0) is the only one that even mentions signalR, but nothing on this topic. I can add a new section to Migration [2.2 to 3.0](https://learn.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-7.0) on SignalR for this topic if that is what you were thinking.

Exactly. That's where it belongs. Put a couple links to that in the more resources section of a couple docs.

**Comment by wadepickett** [comment]
@Rick-Anderson, 

- I moved the note into the include as you suggested.

- I added a small new linkable sub section to the end of the SignalR section of the migration docs.

- I'll assume you still want the Note in the hubs.md topic Brennan called out originally, so it stands out enough to take care of the problem Brennan was concerned with.  Then you said you want "links to [the new migration section] in the more resources section of a couple docs. That sounds like the couple other docs would _only_ have the more resources link, which I provided.

- I added the resource link in the following topics that seemed relevant, (but it is hard for me to tell where you would draw the line at an "onboarding" level on these so hopefully these are right):

[Internal Review - hubcontext.md ](https://review.learn.microsoft.com/en-us/aspnet/core/signalr/hubcontext?view=aspnetcore-8.0&branch=pr-en-us-28398#additional-resources)
[Internal Review - api-design.md ](https://review.learn.microsoft.com/en-us/aspnet/core/signalr/api-design?view=aspnetcore-8.0&branch=pr-en-us-28398#additional-resources)

- I backed out of the other topics.

Thanks for the suggestions.

File: aspnetcore/migration/22-to-30.md [hunk 5]
```diff
@@ -878,6 +878,10 @@ services.AddSignalR(hubOptions =>
 
 In ASP.NET Core 2.2, you could set the `TransportMaxBufferSize` and that would effectively control the maximum message size. In ASP.NET Core 3.0, that option now only controls the maximum size before backpressure is observed.
 
+#### SignalR assemblies in shared framework
+
+ASP.NET Core SignalR assemblies are no longer published to NuGet as packages. Instead, the assemblies are included in the `Microsoft.AspNetCore.App` shared framework, which is installed with the .NET Core SDK and runtime installers. See [Remove obsolete package references](#remove-obsolete-package-references) in this document.
```
Comment (BrennanConroy) [inline aspnetcore/migration/22-to-30.md:883]: Server-side is no longer in packages. The client still comes from packages.

Comment (Rick-Anderson) [inline aspnetcore/migration/22-to-30.md:883]: And then mention what I said previously about `Microsoft.AspNetCore.App` included in newer versions. At this   point, most using this doc will be going to .NET 6 or later and you don't want to waste their time thinking about `Microsoft.AspNetCore.App`

Comment (wadepickett) [reply aspnetcore/migration/22-to-30.md:883]: @Rick-Anderson, thanks. 

Is the change you want something like this, shorter, not mentioning  `Microsoft.AspNetCore.App` and moving up front that server-side assemblies are provided so no action to take?

(Example 1, short, you have the server-side assemblies already installed):
ASP.NET Core SignalR server-side assemblies are now installed with the .NET Core SDK and runtime installers. See [Remove obsolete package references](#remove-obsolete-package-references) in this document.

(Example2, clarifying for your version, you have the server-side assemblies already, then the detail on where it is now. ):
In ASP.NET Core 3.0 or later, SignalR server-side assemblies are installed with the .NET Core SDK and runtime installers. The server assemblies are included in the `Microsoft.AspNetCore.App` shared framework See [Remove obsolete package references](#remove-obsolete-package-references) in this document.

Sorry to have to ask for clarification, but those are my best guesses and what you want.   I looked through the discussion here several times and I was not clear on where you "mention what I said previously about `Microsoft.AspNetCore.App`" and what this text should change to exactly.  I do understand the why.

**Comment by Rick-Anderson** [comment]
ASP.NET Core SignalR server-side assemblies are ~now~ installed with the .NET Core SDK. For more information see `This link in the migration doc`.

**Comment by wadepickett** [comment]
> ASP.NET Core SignalR server-side assemblies are ~now~ installed with the .NET Core SDK. For more information see `This link in the migration doc`.

OK, thanks. Reducing the new section in the migration topic to just that.

File: aspnetcore/migration/22-to-30.md [hunk 6]
```diff
@@ -878,6 +878,10 @@ services.AddSignalR(hubOptions =>
 
 In ASP.NET Core 2.2, you could set the `TransportMaxBufferSize` and that would effectively control the maximum message size. In ASP.NET Core 3.0, that option now only controls the maximum size before backpressure is observed.
 
+#### SignalR assemblies in shared framework
+
+ASP.NET Core SignalR server-side assemblies are installed with the .NET Core SDK. For more information, see [Remove obsolete package references](#remove-obsolete-package-references) in this document.
```
Comment (Rick-Anderson) [inline aspnetcore/migration/22-to-30.md:883]:
```suggestion
ASP.NET Core SignalR server-side assemblies are now installed with the .NET Core SDK. For more information, see [Remove obsolete package references](#remove-obsolete-package-references) in this document.
```

**Review by Rick-Anderson — APPROVED** [review]
OK with me as long as @BrennanConroy  approves

File: aspnetcore/signalr/hubs.md [hunk 7]
```diff
@@ -509,46 +617,62 @@ If you have an exceptional condition you *do* want to propagate to the client, y
 > [!NOTE]
 > SignalR only sends the `Message` property of the exception to the client. The stack trace and other properties on the exception aren't available to the client.
 
+[!INCLUDE[](~/includes/signalr-in-shared-framework.md)]
```
Comment (BrennanConroy) [inline aspnetcore/signalr/hubs.md:620]: This doesn't look good where it is, it's directly after a section that's talking about getting errors on the client (and after another note). Maybe should be after the very first paragraph on this page?

File: aspnetcore/includes/signalr-in-shared-framework.md [hunk 8]
```diff
@@ -0,0 +1,2 @@
+> [!NOTE]
+> ASP.NET Core SignalR doesn't depend on getting many assemblies from NuGet. See [SignalR assemblies in shared framework](xref:migration/22-to-30#signalr-assemblies-in-shared-framework) for more information.
```
Comment (BrennanConroy) [inline aspnetcore/includes/signalr-in-shared-framework.md:2]: Can this be the same text as below?

Comment (wadepickett) [inline aspnetcore/includes/signalr-in-shared-framework.md:2]:
```suggestion
ASP.NET Core SignalR server-side assemblies are now installed with the .NET Core SDK. For more information, see [Remove obsolete package references](#remove-obsolete-package-references) in this document.
```

File: aspnetcore/signalr/hubs.md (hunk 7 above)
Comment (wadepickett) [inline aspnetcore/signalr/hubs.md:620]:
```suggestion
```

Comment (wadepickett) [reply aspnetcore/signalr/hubs.md:620]: @BrennanConroy, Would the end of the first section, ### Configure SignalR hubs be OK with you?   That seems like an appropriate context.

Comment (wadepickett) [reply aspnetcore/signalr/hubs.md:620]: I'll move the note there to the end of the first section ### Configure SignalR hubs,  and if you still want it right after the intro (very first paragraph) instead, that is cool and I will change.

**Review by BrennanConroy — APPROVED** [review]

**Pull request merged**